File: output/GPT3.5-Turbo/xCodeEval_4000/C++/C/1c35e3639a1765497e8b7f0b829606ba.c
#include <stdio.h>
#include <stdlib.h>

typedef long long ll;

#define MOD 998244353

typedef struct {
    ll v;
} mint;

mint make(ll x) {
    mint m;
    m.v = x;
    return m;
}

mint add(mint a, mint b) {
    return make((a.v + b.v) % MOD);
}

mint sub(mint a, mint b) {
    return make((a.v - b.v + MOD) % MOD);
}

mint mul(mint a, mint b) {
    return make((a.v * b.v) % MOD);
}

mint pow(mint a, ll p) {
    mint res = make(1);
    while (p > 0) {
        if (p & 1) {
            res = mul(res, a);
        }
        a = mul(a, a);
        p >>= 1;
    }
    return res;
}

mint inv(mint a) {
    return pow(a, MOD - 2);
}

mint div(mint a, mint b) {
    return mul(a, inv(b));
}

mint fact[2000010];
mint ifact[2000010];

void initFact(int N) {
    fact[0] = make(1);
    for (int i = 1; i <= N; i++) {
        fact[i] = mul(fact[i - 1], make(i));
    }
    ifact[N] = inv(fact[N]);
    for (int i = N - 1; i >= 0; i--) {
        ifact[i] = mul(ifact[i + 1], make(i + 1));
    }
}

mint choose(int a, int b) {
    if (b < 0 || a < b) {
        return make(0);
    }
    return mul(mul(fact[a], ifact[b]), ifact[a - b]);
}

mint solve(ll N) {
    mint res = make(0);
    for (int i = 0; i <= N; i++) {
        if (i == 0) {
            for (int j = 1; j <= N; j++) {
                res = add(res, mul(pow(make(-1), i + j), mul(choose(N, j), pow(make(3), j + N * (N - j)))));
            }
        } else {
            res = add(res, mul(mul(pow(make(-1), i), choose(N, i)), sub(pow(sub(pow(make(3), N - i), make(1)), N), pow(make(3), (N - i) * N)))));
            res = add(res, mul(mul(pow(make(-1), i), choose(N, i)), mul(pow(make(3), i + N * (N - i)), make(3))));
        }
    }
    return sub(make(0), res);
}

int main() {
    ll N;
    scanf("%lld", &N);
    initFact(2000010);
    mint ans = solve(N);
    printf("%lld\n", ans.v);
    return 0;
}

File: output/GPT4o-Mini/xCodeEval_4000/C++/Java/1c35e3639a1765497e8b7f0b829606ba.java
import java.util.*;
import java.io.*;

public class 1c35e3639a1765497e8b7f0b829606ba_nc {
    static final long MOD = 998244353;
    static long[] fact, ifact;

    static long mod(long x) {
        return (x % MOD + MOD) % MOD;
    }

    static long choose(int a, int b) {
        if (b < 0 || a < b) return 0;
        return mod(fact[a] * ifact[b] % MOD * ifact[a - b] % MOD);
    }

    static void initFact(int N) {
        fact = new long[N];
        ifact = new long[N];
        fact[0] = 1;
        for (int i = 1; i < N; i++) {
            fact[i] = mod(fact[i - 1] * i);
        }
        ifact[N - 1] = modInverse(fact[N - 1]);
        for (int i = N - 2; i >= 0; i--) {
            ifact[i] = mod(ifact[i + 1] * (i + 1));
        }
    }

    static long modInverse(long a) {
        return pow(a, MOD - 2);
    }

    static long pow(long a, long p) {
        long res = 1;
        while (p > 0) {
            if ((p & 1) == 1) {
                res = mod(res * a);
            }
            a = mod(a * a);
            p >>= 1;
        }
        return res;
    }

    static long solve(long N) {
        long res = 0;
        for (int i = 0; i <= N; i++) {
            if (i == 0) {
                for (int j = 1; j <= N; j++) {
                    res = mod(res + mod(pow(-1, i + j) * choose(N, j) * pow(3, j + N * (N - j))));
                }
            } else {
                res = mod(res + mod(pow(-1, i) * choose(N, i) * (mod(pow(3, N - i) - 1) * mod(pow(N, N)) - mod(pow(3, (N - i) * N))) * 3));
                res = mod(res + mod(pow(-1, i) * choose(N, i) * mod(pow(3, i + N * (N - i)))));
            }
        }
        return mod(-res);
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int N = scanner.nextInt();
        initFact(2000010);
        System.out.println(solve(N));
    }
}
